soc: stm32f1x: Add support for stop/standby modes

Oleh-Kravchenko opened the pull request:
Add config and overlay to samples for testing stop/standby modes:

- samples/boards/st/power_mgmt/blinky
- samples/boards/st/power_mgmt/wkup_pins

I've measured consumption for each low-power mode:
- stop (regulator in run mode) ~217 uA
- stop (regulator in low-power mode) ~206 uA
- standby mode ~3.5 uA

Low-power mode wakeup timings from the datasheet,
but it barely meets these in reality:
- stop (regulator in run mode) 3.6 us
- stop (regulator in low-power mode) 5.4 us
- standby 50 us

It's possible to use RTC as idle timer to exit from stop mode.

erwango commented on soc/st/stm32/stm32f1x/CMakeLists.txt:
```diff
@@ -1,9 +1,11 @@
+zephyr_sources(soc.c)
+
+if(CONFIG_PM OR CONFIG_POWEROFF)
+    zephyr_library_sources(power.c)
```
It would be better to split in 2 files.
`z_sys_poweroff` is supposed to be available only if `CONFIG_POWEROFF=y` and same thing for PM hooks with `CONFIG_PM=y` having all them in the same file raise the useless issue of wondering what happens in other cases.
Having 2 files as done on other series is much clearer.

erwango commented on soc/st/stm32/stm32f1x/power.c:
```diff
@@ -0,0 +1,113 @@
+{
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+
+#if CONFIG_DEBUG
+	LL_DBGMCU_EnableDBGSleepMode();
+	LL_DBGMCU_EnableDBGStopMode();
+	LL_DBGMCU_EnableDBGStandbyMode();
+#else
+	/* reduce power consumption by disabling Debug Module */
+	LL_DBGMCU_DisableDBGSleepMode();
+	LL_DBGMCU_DisableDBGStopMode();
+	LL_DBGMCU_DisableDBGStandbyMode();
+#endif /* CONFIG_DEBUG */
```
To be put under `soc/st/stm32/common/soc_config.c` as done on other series.

Oleh-Kravchenko replied:
What about? 
```
LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
```
Without it sys_poweroff() and stop0/stop1 modes don't work.

erwango replied:
I would,'t mind having it enabled in soc.c file under `#if defined(CONFIG_PM) || defined(CONFIG_POWEROFF)`.

erwango commented on samples/boards/st/power_mgmt/blinky/boards/nucleo_f103rb.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_COUNTER=y
```
To be put under `soc/st/stm32/stm32f1x/Kconfig.defconfig` so it is set automatically when required:
```
if PM

config COUNTER
	default y

endif
```

erwango reviewed (changes requested):
Some initial comments before we had the possibility to test.

Also, on F4, the following is done to ensure wakeup delay accuracy:
```
config COUNTER_RTC_STM32_SUBSECONDS
	default y if DT_HAS_ST_STM32_RTC_ENABLED
```
Isn't it required in your case ?

Oleh-Kravchenko commented:
> Some initial comments before we had the possibility to test.
> 
> 
> 
> Also, on F4, the following is done to ensure wakeup delay accuracy:
> 
> ```
> 
> config COUNTER_RTC_STM32_SUBSECONDS
> 
> 	default y if DT_HAS_ST_STM32_RTC_ENABLED
> 
> ```
> 
> Isn't it required in your case ?
> 
> 
> 
> 

STM32F1 can count only in seconds.

  depends on !SOC_SERIES_STM32F1X

erwango reviewed (changes requested):
Thanks, code looks ok now, but in order to be sure this will be tested in CI, please update following samples:
samples/boards/st/power_mgmt/wkup_pins/sample.yaml
samples/boards/st/power_mgmt/blinky/sample.yaml

@djiatsaf-st Do you know other tests that should be updated to take this new contribution into account ?

djiatsaf-st commented:
> Thanks, code looks ok now, but in order to be sure this will be tested in CI, please update following samples: samples/boards/st/power_mgmt/wkup_pins/sample.yaml samples/boards/st/power_mgmt/blinky/sample.yaml
> 
> @djiatsaf-st Do you know other tests that should be updated to take this new contribution into account ?

Maybe in the same directory: `power_mgmt/adc`

In the tests folder side, I did not find relevant tests to use.

Oleh-Kravchenko commented:
> > Thanks, code looks ok now, but in order to be sure this will be tested in CI, please update following samples: samples/boards/st/power_mgmt/wkup_pins/sample.yaml samples/boards/st/power_mgmt/blinky/sample.yaml
> > @djiatsaf-st Do you know other tests that should be updated to take this new contribution into account ?
> 
> Maybe in the same directory: `power_mgmt/adc`
> 
> In the tests folder side, I did not find relevant tests to use.

Hello @erwango, @djiatsaf-st,

The LPTIM is not available in STM32F1x, so what the point to edit power_mgmt/adc?

Oleh-Kravchenko committed:
soc: stm32f1x: Add support for stop/standby modes

Add config and overlay to samples for testing stop/standby modes:

- samples/boards/st/power_mgmt/blinky
- samples/boards/st/power_mgmt/wkup_pins

I've measured consumption for each low-power mode:
- stop (regulator in run mode) ~217 uA
- stop (regulator in low-power mode) ~206 uA
- standby mode ~3.5 uA

Low-power mode wakeup timings from the datasheet,
but it barely meets these in reality:
- stop (regulator in run mode) 3.6 us
- stop (regulator in low-power mode) 5.4 us
- standby 50 us

It's possible to use RTC as idle timer to exit from stop mode.

Signed-off-by: Oleh Kravchenko <oleg@kaa.org.ua>

djiatsaf-st commented:
> The LPTIM is not available in STM32F1x, so what is the point to edit power_mgmt/adc?

Hello @Oleh-Kravchenko , You're right, I think it's good. I'll run the CI test to confirm the changes.

djiatsaf-st commented on samples/boards/st/power_mgmt/wkup_pins/sample.yaml:
```diff
@@ -11,5 +11,7 @@ tests:
+      - nucleo_f103rb
     integration_platforms:
       - nucleo_l4r5zi
+      - nucleo_f103rb
```
If  nucleo_f103rb and  nucleo_l4r5zi  use the same gpio driver, then we only need one of them in the integration list
I think you can remove this line.

Oleh-Kravchenko replied:
@djiatsaf-st,
Sorry, I'm little bit confused.
My pull request adds support of LPM for STM32F1x, why GPIO driver means some thing here?

djiatsaf-st replied:
My remark concerns the wkup_pins sample, which manipulates the GPIO, and the integration_platforms properties are conditions to execute this sample.

djiatsaf-st commented on samples/boards/st/power_mgmt/blinky/sample.yaml:
```diff
@@ -18,3 +18,4 @@ tests:
     integration_platforms:
       - nucleo_wb55rg
       - nucleo_f429zi
+      - nucleo_f103rb
```
ditto

djiatsaf-st reviewed (commented):
Otherwise LGTM
